docs/components/rendering_iterables.md

```diff
@@ -1,46 +1,106 @@
     ]
 
 
 def colored_box(color: str):
-    return rx.box(rx.text(color), background_color=color)
+    return rx.button(color, background_color=color)
 
 
-def simple_foreach():
-    return rx.grid(
+def dynamic_buttons():
+    return rx.vstack(
         rx.foreach(IterState.color, colored_box),
-        columns="6",
     )
 
 ```
 
-```md alert warning
-# The type signature of the functions does not matter to the `foreach` component.
+In this first simple example we iterate through a `list` of colors and render a dynamic number of buttons.
+
+The first argument of the `rx.foreach` function is the state var that you want to iterate through. The second argument is a function that takes in an item from the data and returns a component. In this case, the `colored_box` function takes in a color and returns a button with that color.
+
+## For vs Foreach
+
+```md definition
+# Regualar For Loop
+* Use when iterating over constants.
+
+# Foreach
+* Use when iterating over state vars.
+```
+
+The above example could have been written using a regular Python `for` loop, since the data is constant.
+
+```python demo exec
+colors = ["red", "green", "blue"]
+def dynamic_buttons_for():
+    return rx.vstack(
+        [colored_box(color) for color in colors],
+    )
+```
+
+However, as soon as you need the data to be dynamic, you must use `rx.foreach`.
+
+```python demo exec
+class DynamicIterState(rx.State):
+    color: list[str] = [
+        "red",
+        "green",
+        "blue",
+    ]
+
+    def add_color(self, form_data):
+        self.color.append(form_data["color"])
+
+def dynamic_buttons_foreach():
+    return rx.vstack(
+        rx.foreach(DynamicIterState.color, colored_box),
+        rx.form(
+            rx.input(name="color", placeholder="Add a color"),
+            rx.button("Add"),
+            on_submit=DynamicIterState.add_color,
+        )
+    )
+```
+
+## Render Function
+
+The function to render each item can be defined either as a separate function or as a lambda function. In the example below, we define the function `colored_box` separately and pass it to the `rx.foreach` function. 
+
+```python demo exec
+class IterState2(rx.State):
+    color: list[str] = [
+        "red",
+        "green",
+        "blue",
+    ]
+
+def colored_box(color: rx.Var[str]):
+    return rx.button(color, background_color=color)
+
+def dynamic_buttons2():
+    return rx.vstack(
+        rx.foreach(IterState2.color, colored_box),
+    )
 
-Rather the type annotation on the `state var` determines what operations are available (e.g. when nesting).
 ```
 
-## Enumeration
+Notice that the type annotation for the `color` parameter in the `colored_box` function is `rx.Var[str]` (rather than just `str`). This is because the `rx.foreach` function passes the item as a `Var` object, which is a wrapper around the actual value. This is what allows us to compile the frontend without knowing the actual value of the state var (which is only known at runtime).
```

Comment:
maybe a link for the runtime line to: docs/getting-started/basics/#compile-time-vs.-runtime-(important)